source/extensions/quic_listeners/quiche/envoy_quic_server_session.cc

```diff
@@ -0,0 +1,179 @@
+#include "extensions/quic_listeners/quiche/envoy_quic_server_session.h"
+
+#pragma GCC diagnostic push
+// QUICHE allows unused parameters.
+#pragma GCC diagnostic ignored "-Wunused-parameter"
+// QUICHE uses offsetof().
+#pragma GCC diagnostic ignored "-Winvalid-offsetof"
+
+#include "quiche/quic/core/quic_crypto_server_stream.h"
+#pragma GCC diagnostic pop
+
+#include "extensions/quic_listeners/quiche/envoy_quic_server_stream.h"
+
+namespace Envoy {
+namespace Quic {
+
+EnvoyQuicServerSession::EnvoyQuicServerSession(
+    const quic::QuicConfig& config, const quic::ParsedQuicVersionVector& supported_versions,
+    quic::QuicConnection* connection, quic::QuicSession::Visitor* visitor,
+    quic::QuicCryptoServerStream::Helper* helper, const quic::QuicCryptoServerConfig* crypto_config,
+    quic::QuicCompressedCertsCache* compressed_certs_cache, Event::Dispatcher& dispatcher)
+    : quic::QuicServerSessionBase(config, supported_versions, connection, visitor, helper,
+                                  crypto_config, compressed_certs_cache),
+      filter_manager_(*this), dispatcher_(dispatcher), stream_info_(dispatcher.timeSource()) {}
+
+EnvoyQuicServerSession::~EnvoyQuicServerSession() { delete connection(); }
+
+quic::QuicCryptoServerStreamBase* EnvoyQuicServerSession::CreateQuicCryptoServerStream(
+    const quic::QuicCryptoServerConfig* crypto_config,
+    quic::QuicCompressedCertsCache* compressed_certs_cache) {
+  return new quic::QuicCryptoServerStream(crypto_config, compressed_certs_cache, this,
+                                          stream_helper());
+}
+
+quic::QuicSpdyStream* EnvoyQuicServerSession::CreateIncomingStream(quic::QuicStreamId id) {
+  if (!ShouldCreateIncomingStream(id)) {
+    return nullptr;
+  }
+  auto stream = new EnvoyQuicServerStream(id, this, quic::BIDIRECTIONAL);
+  ActivateStream(absl::WrapUnique(stream));
+  setUpRequestDecoder(*stream);
+  return stream;
+}
+
+quic::QuicSpdyStream* EnvoyQuicServerSession::CreateIncomingStream(quic::PendingStream* pending) {
+  auto stream = new EnvoyQuicServerStream(pending, this, quic::BIDIRECTIONAL);
+  ActivateStream(absl::WrapUnique(stream));
+  setUpRequestDecoder(*stream);
+  return stream;
+}
+
+quic::QuicSpdyStream* EnvoyQuicServerSession::CreateOutgoingBidirectionalStream() {
+  // Not allow server initiated stream.
+  NOT_REACHED_GCOVR_EXCL_LINE;
+}
+
+quic::QuicSpdyStream* EnvoyQuicServerSession::CreateOutgoingUnidirectionalStream() {
+  NOT_REACHED_GCOVR_EXCL_LINE;
+}
+
+void EnvoyQuicServerSession::setUpRequestDecoder(EnvoyQuicStream& stream) {
+  ASSERT(http_connection_callbacks_ != nullptr);
+  Http::StreamDecoder& decoder = http_connection_callbacks_->newStream(stream);
+  stream.setDecoder(decoder);
+}
+
+void EnvoyQuicServerSession::OnConnectionClosed(const quic::QuicConnectionCloseFrame& frame,
+                                                quic::ConnectionCloseSource source) {
+  quic::QuicServerSessionBase::OnConnectionClosed(frame, source);
```

Comment:
Somewhere it might be good to document lifetime of various components of the stack.  I have zero confidence we won't have Envoy bits still trying to talk to QUIC bits or QUIC bits trying to talk to Envoy bits right now :-P

Reply:
I don't have much confidence at this point either... But I think in most case, quic session should outlive all the callbacks because they are provided by filters which is owned by FilterManagerImpl which is part of envoy quic session. I was thinking to enhance the life time constraint when these component is wired up with HCM.